### DISCUSSION TITLE
Combobox: should "labelText" be required?

### DISCUSSION THREAD

**Discussion by sebm** [twilio-labs/paste#1188] [Q&A] [answered]
I'm implementing a UI with some comboboxes in it. In its design mocks, there is no label above the combobox; rather, the placeholder text indicates to the user what the field is. I've compensated for this by setting the "aria-label" field. 

My code looks like this.
```
<Combobox
  labelText={''}
  aria-label={'Operator'}
  items={foo)}
  placeholder={'Choose an operator'}
/>
```

Right now, the blank "labelText" is necessary because it's not a nullable prop, but does this fundamentally need to be a required prop?

**Comment by corinne-nagel** [comment]
Hi @sebm , in UI like combobox/input/etc the label above is suggested. If you'd like to see the reasoning for this/examples of what to do [this section on accessibility](https://paste.twilio.design/components/input#accessibility) is helpful.

 We'd also not suggest the label text be the placeholder. Is there a reason the design mocks don't have a label?

**Comment by kylegilbreath** [comment]
Hey @corinne-nagel! So this has come up a few times when migrating some PHP patterns to React, we've run into some places where selects are shown without a label alongside an input. One example is the following:
<img width="663" alt="Screen Shot 2021-02-12 at 3 59 39 PM" src="https://user-images.githubusercontent.com/12417517/107831449-56225f00-6d4b-11eb-833f-e8cddd3f88bd.png">
My suggestion was to do something like this for it:
<img width="879" alt="Screen Shot 2021-02-03 at 12 43 39 PM" src="https://user-images.githubusercontent.com/12417517/107831457-5cb0d680-6d4b-11eb-993f-67f88740f826.png">

But specifically for @sebm example, we are using a set of selects to define a query rule, and allowing the customer to add multiple sets of these. So having the label in for all the selects starts to get really busy and distracting. This is what we quickly came up with. But I would love you thoughts on how to utilize the label still without it getting busy.

<img width="1052" alt="Screen Shot 2021-02-12 at 3 55 27 PM" src="https://user-images.githubusercontent.com/12417517/107831780-18720600-6d4c-11eb-8880-7e4e8b2adca8.png">

For what it's worth, I took inspiration from the Messaging Insights dashboard filters which I know is still not fully migrated to React, but aligned the filters such as this:

<img width="1427" alt="Screen Shot 2021-02-12 at 4 01 58 PM" src="https://user-images.githubusercontent.com/12417517/107831820-350e3e00-6d4c-11eb-97ea-2d54e32035aa.png">

Anyway, so if Labels are required to be included on Inputs, Selects, and Comboboxes, we might want to be explicit about this in the Paste docs for the components. Right now, it's not super clear. From the notes on [Accessibility](https://paste.twilio.design/components/select#accessibility), It seems like although having the Label visually is preferred we still have the option of not visually including the Label but instead using `aria-label` to indicate the label.

**Reply by corinne-nagel** [reply]
Ah I see, thanks for clarifying @kylegilbreath ! 

While it's generally our suggestion that labels be on each select, I agree, it would be really busy with all of those labels. I think for something like this it works to put the label above all of the selects if you're also using the `aria-label`. I think having a label above (similar to MI) helps people understand what the selects are doing, so it still offers some priming/guidance for the select.

Hope that helps!

**Reply by kylegilbreath** [reply]
Cool cool. Thanks for the input @corinne-nagel! I'll work with @sebm to update the designs to have the label shown above the first row of selects, similar to how Message Insights are doing it with filters.

**Reply by kylegilbreath** [reply]
One last question @corinne-nagel or @SiTaggart. If I am only adding visual Labels to the first row of selects (see screenshot), should we label the selects that don't have visible labels directly using `aria-label`, or use the [Screen Reader Only](https://paste.twilio.design/components/screen-reader-only) component to add labels to those selects?

<img width="1293" alt="Screen Shot 2021-02-16 at 11 11 58 AM" src="https://user-images.githubusercontent.com/12417517/108104739-e8c34680-7048-11eb-953d-bef6ad059528.png">

cc: @sebm

**Comment by SiTaggart** [comment, marked as answer]
Hi folks

So minor updates on this. The reason we make Combobox labels required is that it more closely follows the [Input guidelines](https://paste.twilio.design/components/input/#accessibility) and not the Select guidelines because it can take a placeholder, whereas Selects cannot.

Because of that it's very tempting for folks to try and use a Placeholder as a label, which is against the more detailed Input guidelines. 

As for implementation, this has come up before in a different thread but the detail on labels got buried in thread so I don't blame you for not finding it. You can use a screen reader only component in the labelText prop. https://github.com/twilio-labs/paste/discussions/993#discussioncomment-260896

**Reply by kylegilbreath** [reply]
Thanks for the help @SiTaggart !
